src/library_promise.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * @license
+ * Copyright 2023 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+mergeInto(LibraryManager.library, {
+  $promiseMap__deps: ['$handleAllocator'],
+  $promiseMap: "new handleAllocator();",
+
+  $getPromise__deps: ['$promiseMap'],
+  $getPromise: function(id) {
+    return promiseMap.get(id).promise;
+  },
+  emscripten_promise_create__deps: ['$promiseMap'],
+  emscripten_promise_create__sig: 'p',
+  emscripten_promise_create: function() {
+    var promiseInfo = {};
+    promiseInfo.promise = new Promise((resolve, reject) => {
+      promiseInfo.reject = reject;
+      promiseInfo.resolve = resolve;
+    });
+    var id = promiseMap.allocate(promiseInfo);
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_create: ' + id);
+#endif
+    return id;
+  },
+
+  emscripten_promise_destroy__deps: ['$promiseMap'],
+  emscripten_promise_destroy__sig: 'vp',
+  emscripten_promise_destroy: function(id) {
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_destroy: ' + id);
+#endif
+    promiseMap.free(id);
+  },
+
+  emscripten_promise_resolve__deps: ['$promiseMap', '$getPromise'],
+  emscripten_promise_resolve__sig: 'vpip',
+  emscripten_promise_resolve: function(id, result, value) {
+#if RUNTIME_DEBUG
+    err('emscripten_promise_resolve: ' + id);
+#endif
+    var info = promiseMap.get(id);
+    if (result == {{{ cDefine('EM_PROMISE_FULFILL') }}}) {
+      info.resolve(value);
+    } else if (result == {{{ cDefine('EM_PROMISE_MATCH') }}}) {
+      info.resolve(getPromise(value));
+    } else if (result == {{{ cDefine('EM_PROMISE_MATCH_RELEASE') }}}) {
+      info.resolve(getPromise(value));
+      _emscripten_promise_destroy(value);
+    } else if (result == {{{ cDefine('EM_PROMISE_REJECT') }}}) {
+      info.reject(value);
+    } else {
+      abort('invalid promise result');
+    }
+  },
+
+  $makePromiseCallback__deps: ['$callUserCallback',
+                               '$withStackSave',
+                               '$getPromise'],
+  $makePromiseCallback: function(callback, userData) {
+    return (value) => {
+#if RUNTIME_DEBUG
+      dbg('emscripten promise callback: ' + value);
+#endif
+      return withStackSave(() => {
+        var resultPtr = stackAlloc(8);
```

Comment:
Why 8?

Reply:
Ah, I should update this to `POINTER_SIZE`.